[DDP] multiple forward support for static graph

rohan-varma opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #103487
* #103304
* #103283
* #103282

Adds support for multiple forward before bwd call for
static_graph=True.

There are 2 changes:
1) Change tracking of accounting of when to populate static grap related maps
from relying on forward iteration to backward calls
2) In DDP python, don't rely on num_forward iterations == 1 to enqueue the
delay allreduce. Instead use a flag.

Differential Revision: [D46673736](https://our.internmc.facebook.com/intern/diff/D46673736/)

rohan-varma committed:
[DDP] multiple forward support for static graph

Adds support for multiple forward before bwd call for
static_graph=True.

There are 2 changes:
1) Change tracking of accounting of when to populate static grap related maps
from relying on forward iteration to backward calls
2) In DDP python, don't rely on num_forward iterations == 1 to enqueue the
delay allreduce. Instead use a flag.

Differential Revision: [D46673736](https://our.internmc.facebook.com/intern/diff/D46673736/)

[ghstack-poisoned]

rohan-varma committed:
Update on "[DDP] multiple forward support for static graph"


Adds support for multiple forward before bwd call for
static_graph=True.

There are 2 changes:
1) Change tracking of accounting of when to populate static grap related maps
from relying on forward iteration to backward calls
2) In DDP python, don't rely on num_forward iterations == 1 to enqueue the
delay allreduce. Instead use a flag.

Differential Revision: [D46673736](https://our.internmc.facebook.com/intern/diff/D46673736/)

[ghstack-poisoned]

rohan-varma committed:
Update on "[DDP] multiple forward support for static graph"


Adds support for multiple forward before bwd call for
static_graph=True.

There are 2 changes:
1) Change tracking of accounting of when to populate static grap related maps
from relying on forward iteration to backward calls
2) In DDP python, don't rely on num_forward iterations == 1 to enqueue the
delay allreduce. Instead use a flag.

Differential Revision: [D46673736](https://our.internmc.facebook.com/intern/diff/D46673736/)

[ghstack-poisoned]

awgu commented on torch/nn/parallel/distributed.py:
```diff
@@ -2201,6 +2202,7 @@ def _set_static_graph(self):
             )
             return
         self.static_graph = True
+        self.static_graph_delay_allreduce_enqueued = False
```
Should this be public?

rohan-varma replied:
Good point!

awgu commented on torch/csrc/distributed/c10d/reducer.cpp:
```diff
@@ -613,6 +615,10 @@ void Reducer::set_logger(std::weak_ptr<c10d::Logger> logger) {
   std::lock_guard<std::mutex> lock(this->mutex_);
+  if (!first_autograd_hook_called_) {
+    first_autograd_hook_called_ = true;
+    num_bwd_calls_++;
```
Curious: In PyTorch, do people care about `++num_bwd_calls_;` instead of `num_bwd_calls_++`?

rohan-varma replied:
To me here it doesn't make a difference because increment pre/postfix doesn't matter here, right?

awgu replied:
I guess I mean for the minor difference of the temporary created for `num_bwd_calls_++`. I am not sure if our default compiler settings optimize that.

awgu commented on torch/testing/_internal/distributed/distributed_test.py:
```diff
@@ -9769,6 +9769,61 @@ def forward(self, x):
+            torch.cuda.set_device(self.rank)
+            torch.manual_seed(42 << 1337 % (self.rank + 1))
+            model = Net().cuda(self.rank)
+            from copy import deepcopy
```
nit: Do we want this inline?

rohan-varma replied:
nope

awgu commented on torch/testing/_internal/distributed/distributed_test.py:
```diff
@@ -9769,6 +9769,61 @@ def forward(self, x):
+                        loss = a.sum() + b.sum()
+                        loss.backward()
+                    for p in local_model.parameters():
+                        p.grad.data = p.grad / 2
```
IIUC, this `/ 2` is to emulate the division by world size from DDP. Would it be good to divide by the world size explicitly (e.g. for clarity and/or in case someone runs this unit test on >2 GPUs)?

rohan-varma replied:
great point!

awgu reviewed (approved):
This makes sense to me!

rohan-varma committed:
Update on "[DDP] multiple forward support for static graph"


Adds support for multiple forward before bwd call for
static_graph=True.

There are 2 changes:
1) Change tracking of accounting of when to populate static grap related maps
from relying on forward iteration to backward calls
2) In DDP python, don't rely on num_forward iterations == 1 to enqueue the
delay allreduce. Instead use a flag.

Differential Revision: [D46673736](https://our.internmc.facebook.com/intern/diff/D46673736/)

[ghstack-poisoned]

rohan-varma committed:
Update on "[DDP] multiple forward support for static graph"


Adds support for multiple forward before bwd call for
static_graph=True.

There are 2 changes:
1) Change tracking of accounting of when to populate static grap related maps
from relying on forward iteration to backward calls
2) In DDP python, don't rely on num_forward iterations == 1 to enqueue the
delay allreduce. Instead use a flag.

Differential Revision: [D46673736](https://our.internmc.facebook.com/intern/diff/D46673736/)

[ghstack-poisoned]

rohan-varma commented:
@pytorchbot merge -f "CI failure unrelated"

osalpekar commented:
@pytorchbot revert -m "This causes distributed/test_c10d_gloo.py::DistributedDataParallelTest::test_ddp_checkpointing_weight_sharing_use_reentrant_False to fail on trunk starting this commit. See https://github.com/pytorch/pytorch/actions/runs/5269586345/jobs/9528524605 for more details" -c landrace
